pkg/utils/ldap/reconnect.go

```diff
@@ -0,0 +1,341 @@
+// Copyright 2022 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package ldap
+
+// LDAP automatic reconnection mechanism, inspired by:
+// https://gist.github.com/emsearcy/cba3295d1a06d4c432ab4f6173b65e4f#file-ldap_snippet-go
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-ldap/ldap/v3"
+	"github.com/rs/zerolog"
+)
+
+var (
+	errMaxRetries = errors.New("max retries")
+)
+
+type ldapConnection struct {
+	Conn  *ldap.Conn
+	Error error
+}
+
+// ConnWithReconnect maintains an LDAP Connection that automatically reconnects after network errors
+type ConnWithReconnect struct {
+	conn    chan ldapConnection
+	reset   chan *ldap.Conn
+	retries int
+	logger  *zerolog.Logger
+}
+
+// Config holds the basic configuration of the LDAP Connection
+type Config struct {
+	URI          string
+	BindDN       string
+	BindPassword string
+	TLSConfig    *tls.Config
+}
+
+// NewLDAPWithReconnect Returns a new ConnWithReconnect initialized from config
+func NewLDAPWithReconnect(config Config) *ConnWithReconnect {
+	conn := ConnWithReconnect{
+		conn:    make(chan ldapConnection),
+		reset:   make(chan *ldap.Conn),
+		retries: 1,
+	}
+	logger := zerolog.Nop()
+	conn.logger = &logger
+	go conn.ldapAutoConnect(config)
+	return &conn
+}
+
+// SetLogger sets the logger for the current instance
+func (c *ConnWithReconnect) SetLogger(logger *zerolog.Logger) {
+	c.logger = logger
+}
+
+// Search implements the ldap.Client interface
+func (c *ConnWithReconnect) Search(sr *ldap.SearchRequest) (*ldap.SearchResult, error) {
+	conn, err := c.getConnection()
+
+	if err != nil {
+		return nil, err
+	}
+
+	var res *ldap.SearchResult
+	for try := 0; try <= c.retries; try++ {
```

Comment:
1) Can we move this loop to a dedicated function? It is used by all Methods so generates lots of duplication.

2) Should we do a more sophisticated retry logic? For example `backoff.Exponential` proved valuable in my experience,

Reply:
I added a separate retry function now.  But I'd rather move a more sophisticated retry mechanism to a later PR if that's ok